File: SplitArrayLargestSum.java
class SplitArrayLargestSum {
    public int splitArray(int[] nums, int m) {
        int sum = 0, max = 0;
        for (int num : nums) {
            sum += num;
            if (num > max)
                max = num;
        }

        int l = max, r = sum;
        int res = r;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (canSplit(mid, nums, m)) {
                res = mid;
                r = mid - 1;
            } else
                l = mid + 1;
        }

        return res;
    }

    private boolean canSplit(int largest, int[] nums, int m) {
        int subarray = 0;
        int curSum = 0;

        for (int num : nums) {
            curSum += num;
            if (curSum > largest) {
                subarray++;
                curSum = num;
            }
        }
        return subarray + 1 <= m;
    }
}

// Time: O(N * log(sum of all elements))
// Space: O(1)

File: split-array-largest-sum.js
var splitArray = (nums, m) => {
  return dfs(0, m);

  function dfs(i, m, memo = {}) {
    if (m === 1) {
      let sum = 0;
      for (let pos = i; pos < nums.length; pos++) sum += nums[pos];
      return sum;
    }
    if ([i, m] in memo) return memo[[i, m]];

    let res = Number.POSITIVE_INFINITY,
      curSum = 0;
    for (let j = i; j < nums.length - m + 1; j++) {
      curSum += nums[j];
      const maxSum = Math.max(curSum, dfs(j + 1, m - 1, memo));
      res = Math.min(res, maxSum);
      if (curSum > res) break;
    }

    memo[[i, m]] = res;
    return res;
  }
};

console.log(splitArray([7, 2, 5, 10, 8], 2)); // logs 18

// Time: O(N^2 * M)
